### PR TITLE
LPS-65105

### PR DESCRIPTION
**Pull request by juangon**
Hi @brunobasto. This issue can confuse users when using SPA as requests are being cached. Hope it's ok for you too.

### CODE REVIEW

**Comment by brunobasto** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by jbalsas** [comment]
Hey guys, I just sumbled upon this. I'm not so sure about this and I don't think the related ticket is a common enough use case to grant disabling our client cache. This feature gives us a huge boost in perceived performance, so we shouldn't just discard it without further consideration.

**Comment by brunobasto** [comment]
Hey Chema, I've been thinking about this too and I think Juan's fix maybe correct for now. The problem is that we have no cache invalidation policy whatsoever, resulting on the behavior described by the ticket.

Now that we rarely refresh the page, it may occur that one user remains if the page open for several days and will not see updates made to a website unless he manually hit crtl + r or click on the browser refresh icon.

it's good to remember that we have etag filter. If properly configured, the requests to unchanged pages should be quite fast.

Another option would be to have local cache enabled and an open socket with the server so that we know when something have changed and clear the local cache. (just brainstorming here).

**Comment by jbalsas** [comment]
Let's think this through, then. This effectively wipes out the client cache entirely. That means that we are degrading everyone's performance to cover an edge case.

If this is a valid use case, then we could time-tag the cache entries on the client side and invalidate them based on a time-to-live setting.

**Comment by juangon** [comment]
@jbalsas I think is not an edge case at all. Everyone would click on Page links, and they would get stale data if they had clicked before.

Additionally, I am fixing https://issues.liferay.com/browse/LPS-63235 which would solve the server cache that wasn't working for some time.

**Comment by juangon** [comment]
BTW people doesn't have to know how to refresh a page (F5) or if Liferay is using SPA or not. Users would just use the links, as that is the way they should behave.

If you are seeing bad performance in portal, we should investigate why, but caching pages in client without having a cache expiration should not be the solution for those problems (@brunobasto idea maybe great, because we won't ever have stale data but caching data in browser at the same time).

**Comment by jbalsas** [comment]
How often do those "pages" get new updates? For how long does a regular user keep a tab open in the browser and doesn't close the app? My feeling is that what you're describing is an edge case, but of course, everything can be debated.

In any case, server cache has a different purpose than client cache. We are testing the portal in a local loop, but in high latency networks, the cost of simply doing a request to the server to get the same content you already have is quite high. And could be avoided.

Again, I'd suggest to find out what's a good compromise for client cache time-to-live and tag the cache on the client side. If you update your website once a month, it doesn't make sense to discard this. If you update it every hour, then you can tweak it to adapt it to your needs.

**Comment by juangon** [comment]
But we shouldn't take that decision by ourselves. What do you mean by "tweak it to adapt it to your needs"? Is there a way to tweak this?

**Comment by jbalsas** [comment]
That's what I'm proposing. 
- Add a `cacheTimeToLive` attribute. 
- By default, we can set it to `Infinity`, which would be pretty much the behaviour you're seeing. This is good for most of the websites that don't have a high amount of updates. Please, think again in how you use the browser. How often do you keep a tab open and the browser app open.
- If you're site updates at a very high pace, then you can set the attribute to `0`, which will effectively turn client cache off and things will behave as in your proposed solution.
- In between, you can set it to... `1day`, `1week`, as that may be the pace of the updates of your site. Again, keep in mind that if you update a site weekly, a user would need to have that tab open and its browser for a week. I still consider this an edge case that shouldn't affect a regular user that navigates to your site, does its thing and leaves.

**Comment by juangon** [comment]
@jbalsas I see your solution could be right but, we already deal with cache in server side. Pages are cached unless some new layout is added, portletpreference is changed or any journal content is updated.

Why do we need another cache layer?

**Comment by juangon** [comment]
What I want to say is that, having a server side cache, we don't need any cache time to live. Cache is only refreshed when needed. No "magic" numbers.

**Comment by juangon** [comment]
That said, your new feature for client side tunning could be great for some cases, but I would add this as disabled and document it in dev.liferay.com as a performance tweak.

**Comment by jbalsas** [comment]
@juangon, we need a client cache if we care about the perceived performance of your site. The server cache is there to protect the server from unnecessary processing. This, however, doesn't save the user from unnecessary network requests.

We'd still be doing an unnecessary request. In a cache hit scenario, the network latency becomes the degrading factor, so you'd be waiting at least 200-300ms of RTT just to get some content you already had. In addition, you're incurring in extra bandwidth (and battery) waste.

My suggestion is leave this as is (which is to me the sensible, not magic, default), and implement the cache expiration policy either in portal or directly in Senna if Eduardo agrees with it.

**Comment by juangon** [comment]
@jbalsas sorry I disagree about letting as is. Users are used to previous behavior (all contents are fresh) , hence this will cause a lot of issues.
If you want, add this improvement, but disabled by default. We could add this as an improvement, but users should read first the consequences of enabling this behavior.

**Comment by brunobasto** [comment]
I tend to agree with Juan. At least until we have a better cache expiration policy. What concerns me is dynamic content. Adding portlets may not be something that happens every hour, but having one portlet with dynamic content in it is (asset publisher?).

**Comment by juangon** [comment]
Yes, @brunobasto  we have many portlets with dynamic contents. In fact, I had to disable the layout-cacheable enabled in the mentioned LPS for cached pages, because some of them were showing stale data too (it wasn't before because Cache wasn't working). I think that is not the way it should be for a portal showing dynamic content like ours.

**Comment by jbalsas** [comment]
@juangon, following that argument, we should ship SPA completely disabled by default. Users should be aware of the consequences before enabling this behaviour. 

I agree with the need of properly documenting everything and informing users, but we do need to take decisions (and we do it everyday) based on the general performance of the platform and the most common use case, and again, I don't think this is it. We also don't have all the data, so feels like we are willing to take this decision lightly without all the information at hand (it may turn out that our clients all want this disabled by default, but we don't know right now).

If what we need is a better (or at least an existing) cache expiration policy, let's build it.

In any case, if you two guys agree, I'd say you should mention it to Eduardo as well and go for it if he's onboard.

**Comment by brunobasto** [comment]
@eduardolundgren

Can you share your thoughts?

**Comment by juangon** [comment]
Well, I work in Support and usually know what users want or think. They usually don't want to change behavior from one version or another. The policy to follow  between versions is letting users enable an improvement if that improvement is going to change the previous behavior. But not enable is as default if that would cause issues to customers from previous versions. In this case issues are very noticeable comparing to previous one.
Additionally, I thought SPA enabled had more advantages even without this cache (for example, resources aren't requested again etc). So I think SPA without cache is perfectly fine and users will gain performance for sure.

That said, until that cache policy is built, I would disable for now as DXP would go out soon and IMHO this will cause issues.

**Comment by jbalsas** [comment]
I think a cache policy could have been built in in Sennajs in the combined time we've spent discussing this here 😓.

Again, I think we're seeing this from our testing point of view and not with how real people interact with a site in real life. This probably would hardly ever be a problem to a real user navigating to a site, doing some stuff and leaving. In return, he'll have a way better perceived performance.

Be as it may, I think I've exposed my concerns. You guys can decide how to move forward with this.

**Comment by eduardolundgren** [comment]
I highly recommend to ship Liferay with SPA **enabled**.

Regarding the cache, it is possible to define a good cache expiration policy, e.g. after the user submit a form, or perform an ajax to an action url.

Disabling the SPA cache is the easier solution, if you want to go with the easier solution, this is the way to go, although I recommend to define a policy instead.

**Comment by brunobasto** [comment]
@eduardolundgren 

We already clear the cache whenever the user submits a form or perform an ajax. However, if **user A** has a cached page and **user B** performs an action that updates that page, then we don't know that we should clear the cache for **user A**

**Comment by juangon** [comment]
Ok so, as @eduardolundgren said, we could go to the easy fix: disable cache for now (hence merge with pull). Then, create a new LPS (Story type) and there discuss the possibilities of this cache (time to leave, request to server to know if layout need a page refresh, etc). For this new story LPS I am ok with the options as long as: 1-All different behavior is disabled by default. 2-It can be configured easily by users.

Thanks!

**Comment by jbalsas** [comment]
I don't think I interpreted @eduardolundgren's comment as a "go for it" as @juangon did. And again, I'd rather not merge this in, but define the policy and have it in place before the DXP release.

**Comment by jbalsas** [comment]
We may need to bring in someone else 😞

**Comment by juangon** [comment]
Ok @jbalsas. As long as it is implemented for DXP and is disabled by default, I am ok with it.

I can bring in Support mates for discussion too.

**Comment by jbalsas** [comment]
You know I'm not sold on the "disabled by default" feature too 😉

**Comment by juangon** [comment]
Sorry @jbalsas but that's the way these "improvements" work when upgrading. Policy is to disable new behaviors that users won't expect. Improvements changing behaviors (like not seeing fresh content) should be disabled by default.

**Comment by brunobasto** [comment]
@juangon But if we solve the cache invalidation policy, then it can be enabled by default, right?

**Comment by juangon** [comment]
@brunobasto it depends what kind of policy we are following. If we, for example, make a request to server to know if a page should be refreshed or not, then it could be, as we won't change previous behavior.
If that policy is based on some "magical" numbers about time to live, I would disable by default too, as that number would depend on customer environment that should be tweaked by them.

**Comment by jbalsas** [comment]
I think that if we were to follow _Policy is to disable new behaviors that users won't expect. Improvements changing behaviors (like not seeing fresh content) should be disabled by default._, then SPA should be completely disabled by default. 

If people want this feature, then they can enable knowing the consequences (such as client cache). 

How about we get clarification about this?

**Comment by juangon** [comment]
Sorry about my lack of knowledge on this @jbalsas, but isn't any other advantage on using SPA even with cache disabled?

About clarification, I asked for more opinions to Support coworkers...

**Comment by jbalsas** [comment]
Hey @juangon, there is, but as you mention, it also changes the behaviour. If there is such a policy and we were to follow it, I think it should be done alltogether.

**Comment by brunobasto** [comment]
Closing as cache invalidation option is implemented as discussed here: https://github.com/brianchandotcom/liferay-portal/pull/38948
